ark/phenotyping/cluster.py

```diff
@@ -0,0 +1,151 @@
+           x_mesh, y_mesh, t, num_iters):
+    """Updates the weights, learning rate, and sigma parameters
+
+    Args:
+        sample (numpy.ndarray):
+            A row in the pixel matrix identifying information associated with one pixel
+        weights (numpy.ndarray):
+            A square matrix identifying the weights between two neurons
+        winning_coords (tuple):
+            A coordinate array which indicates the winning neuron's position
+        sigma (float):
+            Determines the spread of the Gaussian neighborhood function
+        learning_rate (float):
+            Determines how sensitive the weight updates will be to new data
+        x_mesh (numpy.ndarray):
+            The x coordinate matrix of the weights vectors
+        y_mesh (numpy.ndarray):
+            The y coordinate matrix of the weights vectors
+        t (int):
+            The current iteration index
+        num_iters (int):
+            The maximum number of iterations
+
+    Returns:
+        numpy.ndarray:
+            The updated weights matrix
+    """
+
+    # update learning rate with asymptotic decay
+    decay_lr = learning_rate / (1 + t / (num_iters / 2))
+
+    # update sigma with asymptotic decay
+    decay_sig = sigma / (1 + t / (num_iters / 2))
+
+    # return a Gaussian centered around the winning coordinates
+    d = 2 * np.pi * decay_sig * decay_sig
+    ax = np.exp(-np.power(x_mesh - x_mesh.T[winning_coords], 2) / d)
+    ay = np.exp(-np.power(y_mesh - y_mesh.T[winning_coords], 2) / d)
+    g = (ax * ay).T * decay_lr
+
+    # update the weights based on the Gaussian neighborhood
+    weights += np.einsum('ij, ijk->ijk', g, sample - weights)
+
+    return weights
+
+
+def train_flowsom(pixel_mat, x_neurons, y_neurons, num_iters,
+                  sigma=1.0, learning_rate=0.5, random_seed=0):
+    """Trains the SOM by iterating through the each data point and updating the params
+
+    Args:
+        pixel_mat (pandas.DataFrame):
+            A matrix with pixel-level channel information for non-zero pixels in img_xr
+        x_neurons (int):
+            The number of x neurons to use
+        y_neurons (int):
+            The number of y neurons to use
+        num_iters (int):
+            The maximum number of iterations for training
+        sigma (float):
+            Determines the spread of the Gaussian neighborhood function
+        learning_rate (float):
+            Determines how sensitive the weight updates will be to new data
+        random_seed (int):
+            The seed to set for random weight initialization
+
+    Returns:
+        numpy.ndarray:
+            The weights matrix after training
+    """
+
+    # define the random generator
+    rand_gen = np.random.RandomState(random_seed)
+
+    # initialize the weights and normalize
+    weights = rand_gen.rand(x_neurons, y_neurons, pixel_mat.shape[1]) * 2 - 1
+    weights /= np.linalg.norm(weights, axis=-1, keepdims=True)
```

Comment:
This is just to make sure all weights are in [0, 1] right?

Reply:
Yup, we can remove if it's not necessary, just wanted to make sure we were consistently following MiniSOM's logic.

Reply:
No it's definitely useful, just wanted to make sure I understood what was going on